### PR TITLE
vsr: drop request with non-zero timestamps instead of crashing

### PR DESCRIPTION
**Pull request by matklad**
This assert can trip if the client sends us bad data. Thus, a single bit flip in a non ECC-protected (or outright malicious) client can kill the leader.

### CODE REVIEW

File: src/state_machine.zig [hunk 1]
```diff
@@ -188,18 +190,20 @@ pub fn StateMachineType(comptime Storage: type, comptime constants_: struct {
...
-            var sum_reserved_timestamps: usize = 0;
+        ) !void {
+            var input_timestamp_or: usize = 0;
+            var prepare_timestamp = self.prepare_timestamp;
             var events = mem.bytesAsSlice(Event(operation), input);
             for (events) |*event| {
-                sum_reserved_timestamps += event.timestamp;
-                self.prepare_timestamp += 1;
-                event.timestamp = self.prepare_timestamp;
+                input_timestamp_or |= event.timestamp;
```
Comment (matklad) [inline src/state_machine.zig:198]: changed this to `|` to not worry about overflows.

File: src/vsr/replica.zig [hunk 2]
```diff
@@ -835,7 +835,12 @@ pub fn ReplicaType(
             const prepare_timestamp = self.state_machine.prepare(
                 message.header.operation.cast(StateMachine),
                 message.body(),
-            );
+            ) catch |err| switch (err) {
+                error.NonZeroTimestamp => {
+                    log.err("{}: on_request: dropping (non zero timestamp)", .{self.replica});
```
Comment (sentientwaffle) [inline src/vsr/replica.zig]: Additionally log the `message.header.client` and the `message.header.operation.cast(StateMachine)` — they could help pinpoint the origin.

**Comment by matklad** [comment]
Not sure what's the best way to test this. It seems that `./zig/zig build test` doesen't even compile the relevant call-site in replica.zig :fearful:

**Comment by sentientwaffle** [comment]
Yeah, the replica is really only tested by the VOPR (simulator) right now. I'm planning to improve that incrementally, but it is a pretty big task. :confused:  Also, when the VOPR is built with `mode=ReleaseSafe` log lines aren't compiled, which is something of a gotcha (use `Debug`).

Also the VOPR currently has some spurious failures due to state transfer not being implemented.

File: src/state_machine.zig [hunk 3]
```diff
@@ -188,18 +190,20 @@ pub fn StateMachineType(comptime Storage: type, comptime constants_: struct {
             self: *StateMachine,
             comptime operation: Operation,
             input: []u8,
-        ) void {
-            var sum_reserved_timestamps: usize = 0;
+        ) !void {
+            var input_timestamp_or: usize = 0;
```
Comment (sentientwaffle) [inline src/state_machine.zig]: This should be a `u64`. Unfortunately, unlike Rust, Zig won't complain about a mismatched `usize` vs `u64` like this unless you actually build for 32-bit.

**Comment by sentientwaffle** [comment]
Also, the part in the state machine can be tested at least (verify that the error is returned).

File: src/state_machine.zig (hunk 3 above)
Comment (matklad) [reply src/state_machine.zig]: excellent catch, thanks!

**Comment by matklad** [comment]
> Also, the part in the state machine can be tested at least (verify that the error is returned).

Yeah, I'd be a bit reluctant to add such test:

* it doesn't test what we want (eg, the callsite can do `catch unreachabel // YOLO`)
* it'll freeze the API a bit and make it harder to refactor it. Eg, we can imagine moving this check to the code which verifies the checksum, and then we'll have to refactor the unit-test
* there's "negative" value in having a lot of "inappropriate" tests, as this gives a false sense of security. 

(to make sure that the argument doesn't prove too much: it does seem correct to put state-transition tests here, as this is the core of the state-machine. This prepare stuff though is exactly about interface between components)

**Comment by sentientwaffle** [comment]
Closing; I'm going to take this over & use a different approach which will make invalid timestamps into regular (per-account/transfer) error codes.

**Pull request closed without merging**